pkg/client/k8s/v1alpha1/unstructured.go

```diff
@@ -218,45 +216,137 @@ func UpdateLabels(o UnstructuredOptions) UnstructuredMiddleware {
 	}
 }
 
-// UnstructuredUpdater updates an unstructured instance by executing all the
-// provided updaters
-func UnstructuredUpdater(updaters []UnstructuredMiddleware) UnstructuredMiddleware {
-	return func(given *unstructured.Unstructured) *unstructured.Unstructured {
-		for _, u := range updaters {
-			given = u(given)
-		}
+// UnstructuredMiddlewareList is a type definition of a list of unstructured
+// middlewares
+type UnstructuredMiddlewareList []UnstructuredMiddleware
+
+// Update updates the given unstructured instance against all the available
+// middlewares
+func (l UnstructuredMiddlewareList) Update(given *unstructured.Unstructured) (updated *unstructured.Unstructured) {
+	if len(l) == 0 || given == nil {
 		return given
 	}
+	for _, middleware := range l {
+		if middleware == nil {
+			continue
+		}
+		given = middleware(given)
+	}
+	return given
 }
 
-// UnstructList represents a list of unstructured instances
-type UnstructList struct {
-	Items []*unstructured.Unstructured
+// PredicateListOp represents the type of operation categorizing a given list of
+// predicates
+type PredicateListOp string
+
+const (
+	// AllPredicates represents AND logical operator
+	AllPredicates PredicateListOp = "and"
+	// AnyPredicates represents OR logical operator
+	AnyPredicates PredicateListOp = "or"
+	// NotPredicates represents NOT logical operator
+	NotPredicates PredicateListOp = "not"
+)
+
+// UnstructPredicateList represents a list of predicates with a logical operator
+// to evaluate the condition
+type UnstructPredicateList struct {
+	Op    PredicateListOp
+	Items map[string]UnstructuredPredicate
+}
+
+func (l UnstructPredicateList) String() string {
+	var names []string
+	for name := range l.Items {
+		names = append(names, name)
+	}
+	return fmt.Sprintf("unstruct predicate list: op='%s', items='%s'", l.Op, strings.Join(names, ","))
+}
+
+// isNot does a NOT logic among the predicate list against the given
+// unstructured instance
+func (l UnstructPredicateList) isNot(u *unstructured.Unstructured) bool {
```

Comment:
NOT is a unary operator whereas we are supplying a list of operands to operate on. How is this logic used?